### PR TITLE
Add ability to decode\encode stored values and to force cache refresh

### PR DESCRIPTION
**Pull request by adi-rose**

### CODE REVIEW

File: gamla/higher_order.py [hunk 1]
```diff
@@ -73,12 +76,17 @@ def decorator(f: Callable):
...
                     ignore_first_arg(make_key),
-                    ignore_first_arg(f),
+                    functional_generic.compose_left(ignore_first_arg(f), encode),
                 ),
                 functional_generic.side_effect(functional_generic.star(set_item)),
                 operator.second,
+                decode,
+            ),
+            functional_generic.ternary(
+                operator.identity(construct.just(force)),
```
Comment (JonSc8) [inline gamla/higher_order.py:86]: I think you can remove the gamla.identity

File: gamla/higher_order_test.py [hunk 2]
```diff
@@ -117,3 +118,57 @@ async def f(x):
     )(f)("something")
```
Comment (JonSc8) [inline gamla/higher_order_test.py:118]: we need to call this function one more time to assert the get_item

File: gamla/higher_order_test.py [hunk 3]
```diff
@@ -117,3 +118,57 @@ async def f(x):
...
+    d = {}
+
+    def get_item(key: str):
+        return d[key]
+
+    def set_item(key: str, value):
+        d[key] = value
+        return
+
+    @assert_max_called(1)
```
Comment (JonSc8) [inline gamla/higher_order_test.py:133]: assert called twice since force is true (and actually call it twice)

File: gamla/higher_order_test.py [hunk 4]
```diff
@@ -85,14 +85,14 @@ def set_item(key: str, value):
...
-            set_item,
-            functional.make_hashed_call_key,
-        )(f)("something")
-        == "something"
-    )
+    cached_function = higher_order.persistent_cache(
+        get_item, set_item, functional.make_hashed_call_key
+    )(f)
+
+    result_set = cached_function("something")
```
Comment (JonSc8) [inline gamla/higher_order_test.py:92]: no need to set a variable here, you can just call `cached_function("something")`

File: gamla/higher_order_test.py [hunk 5]
```diff
@@ -110,10 +110,68 @@ async def set_item(key: str, value):
...
-        get_item,
-        set_item,
-        functional.make_hashed_call_key,
-    )(f)("something")
-    assert result == "something"
+    cached_function = higher_order.persistent_cache(
+        get_item, set_item, functional.make_hashed_call_key
+    )(f)
+
+    result_set = await cached_function("something")
```
Comment (JonSc8) [inline gamla/higher_order_test.py:117]: same

File: gamla/higher_order_test.py [hunk 6]
```diff
@@ -85,14 +85,14 @@ def set_item(key: str, value):
...
-    )
+    cached_function = higher_order.persistent_cache(
+        get_item, set_item, functional.make_hashed_call_key
+    )(f)
+
+    result_set = cached_function("something")
+    result_get = cached_function("something")
+    assert result_set == "something"
+    assert result_get == "something"
     assert d == {"c3aa999f887e4eb8a1dda68862dcf172a78b5d30": "something"}
```
Comment (JonSc8) [inline gamla/higher_order_test.py:96]: extract the dict to a private constant as it is used multiple times

File: gamla/higher_order_test.py [hunk 7]
```diff
@@ -85,14 +85,14 @@ def set_item(key: str, value):
...
-        == "something"
-    )
+    cached_function = higher_order.persistent_cache(
+        get_item, set_item, functional.make_hashed_call_key
+    )(f)
+
+    result_set = cached_function("something")
+    result_get = cached_function("something")
+    assert result_set == "something"
+    assert result_get == "something"
```
Comment (JonSc8) [inline gamla/higher_order_test.py:95]: also extract "something" to a constant

File: gamla/higher_order_test.py [hunk 8]
```diff
@@ -110,10 +110,68 @@ async def set_item(key: str, value):
...
+def test_persistent_cache_force():
+    d = {}
+
+    def get_item(key: str):
+        return d[key]
+
+    def set_item(key: str, value):
+        d[key] = value
+        return
+
```
Comment (JonSc8) [inline gamla/higher_order_test.py:133]: Can we somehow extract this so we won't duplicate this once for each test?

File: gamla/higher_order.py [hunk 9]
```diff
@@ -64,6 +64,9 @@ def persistent_cache(
     get_item: Callable[[str], Any],
     set_item: Callable[[str, Any], None],
     make_key: Callable[[Any], str],
+    encode: Callable[[Any], Any],
+    decode: Callable[[Any], Any],
+    force: bool,
 ) -> Callable:
     """Wraps a function with persistent cache. Gets the item getter and item setter as parameters."""
```
Comment (nitzo) [inline gamla/higher_order.py:71]: please add the different parameters with explanation to the doc string

File: gamla/higher_order_test.py [hunk 10]
```diff
@@ -1,7 +1,13 @@
 import asyncio
 import functools
+import zlib
```
Comment (nitzo) [inline gamla/higher_order_test.py:3]: do we need to add this to the dependencies (setup.py)

**Review by nitzo — APPROVED** [review]

File: gamla/higher_order_test.py (hunk 10 above)
Comment (adi-rose) [reply gamla/higher_order_test.py:3]: Yes it was not there - added it now

Comment (adi-rose) [reply gamla/higher_order_test.py:3]: sorry - on second look seems it is a built in library and does not need to be in setup

**Pull request merged**